internal/controller/streams.go

```diff
@@ -197,105 +167,289 @@ func (sm *streamManager) BootUpStreamAndReconnect(ctx context.Context, cancel co
+				return err
 			}
-		}()
+		}
 	}
 	return nil
 }
 
-// ExponentialStreamConnect will continue to reboot and restart the main operations within the operator if any disconnects or errors occur.
-func ExponentialStreamConnect(ctx context.Context, logger *zap.SugaredLogger, envMap EnvironmentConfig, bufferedGrpcSyncer *BufferedGrpcWriteSyncer) {
-	var backoff = 1 * time.Second
-	authn := Authenticator{Logger: logger}
+// connectAndStreamCiliumNetworkFlows creates ciliumNetworkFlows client and begins the streaming of network flows.
+func connectAndStreamCiliumNetworkFlows(logger *zap.SugaredLogger, sm *streamManager) error {
+	ciliumCtx, ciliumCancel := context.WithCancel(context.Background())
+	defer ciliumCancel()
+
+	sendCiliumNetworkFlowsStream, err := sm.streamClient.client.SendKubernetesNetworkFlows(ciliumCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.networkFlowsStream = sendCiliumNetworkFlowsStream
+
+	err = sm.StreamCiliumNetworkFlows(ciliumCtx, sm.streamClient.ciliumNamespace)
+	if err != nil {
+		if errors.Is(err, ErrHubbleNotFound) || errors.Is(err, ErrNoPortsAvailable) {
+			logger.Warnw("Disabling Cilium flow collection", "error", err)
+			return ErrStopRetries
+		}
+		return err
+	}
+
+	return nil
+}
+
+// connectAndStreamResources creates resourceStream client and begins the streaming of resources.
+func connectAndStreamResources(logger *zap.SugaredLogger, sm *streamManager) error {
+	resourceCtx, resourceCancel := context.WithCancel(context.Background())
+	defer resourceCancel()
+
+	SendKubernetesResourcesStream, err := sm.streamClient.client.SendKubernetesResources(resourceCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.resourceStream = SendKubernetesResourcesStream
+
+	err = sm.StreamResources(resourceCtx, resourceCancel)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream resources", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// connectAndStreamLogs creates sendLogs client and begins the streaming of logs.
+func connectAndStreamLogs(logger *zap.SugaredLogger, sm *streamManager) error {
+	logCtx, logCancel := context.WithCancel(context.Background())
+	defer logCancel()
+
+	SendLogsStream, err := sm.streamClient.client.SendLogs(logCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.logStream = SendLogsStream
+	sm.bufferedGrpcSyncer.UpdateClient(sm.streamClient.logStream, sm.streamClient.conn)
+
+	err = sm.StreamLogs(logCtx)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream logs", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// Generic function to manage any stream with backoff and reconnection logic.
+func manageStream(logger *zap.SugaredLogger, connectAndStream func(*zap.SugaredLogger, *streamManager) error, sm *streamManager, done chan struct{}) {
```

Comment:
You close `done` before every `return`.
You can as well do that in a `defer`.
```suggestion
func manageStream(logger *zap.SugaredLogger, connectAndStream func(*zap.SugaredLogger, *streamManager) error, sm *streamManager, done chan struct{}) {
	defer close(done)
```